File: Data Structures & Algorithms/valid-sudoku/submission-1.rs
impl Solution {
pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
    let mut rows = vec![0; 9];
    let mut cols = vec![0; 9];
    let mut cells = vec![0; 9];
    for x in 0..9 {
        for y in 0..9 {
            if board[x][y] == '.' {
                continue;
            }
            let value = board[x][y] as i32 - '1' as i32;
            let bits = 1 << value;
            let cell_id = (x / 3) * 3 + (y / 3);
            if (rows[x] & bits) != 0 || (cols[y] & bits) != 0 || (cells[cell_id] & bits) != 0 {
                return false;
            }
            rows[x] |= bits;
            cols[y] |= bits;
            cells[cell_id] |= bits;
        }
    }
    true
}
}

File: Data Structures & Algorithms/search-for-word/submission-1.cpp
class Solution {
 private:
  int rows, cols;

  bool dfs(vector<vector<char>>& board, string& word, int i, int j, int idx) {
    if (idx == word.size()) return true;
    if (i < 0 || j < 0 || i >= rows || j >= cols || board[i][j] != word[idx] ||
        board[i][j] == '#') {
      return false;
    }
    board[i][j] = '#';
    bool res = dfs(board, word, i, j + 1, idx + 1) ||
               dfs(board, word, i, j - 1, idx + 1) ||
               dfs(board, word, i + 1, j, idx + 1) ||
               dfs(board, word, i - 1, j, idx + 1);
    board[i][j] = word[idx];
    return res;
  }

 public:
  bool exist(vector<vector<char>>& board, string word) {
    rows = board.size();
    cols = board[0].size();
    for (int i{}; i < rows; ++i) {
      for (int j{}; j < cols; ++j) {
        if (board[i][j] == word[0] && dfs(board, word, i, j, 0)) {
          return true;
        }
      }
    }
    return false;
  }
};
